README.md
# ReBooth

## Disclaimer

This code is still full of commented parts and debugging instructions and needs to be cleaned up. Also, some of the internal documentation is still in Italian and needs to be translated. Use it at your own risk. We do not take any responsibility for any kind of damage it might cause. That said, once you meet the prerequisites and edit the configuration files accordingly, ReBooth should work as is. Please, read the License section below before downloading ReBooth.

## What is ReBooth and how it works

ReBooth (which stands for _Remote Booth_) is a [WebRTC](https://webrtc.org/) based platform for conference interpreter training, conceived and developed for the Department of Interpretation and Translation (DIT) at the University of Bologna by Gabriele Carioli and Nicoletta Spinolo.

The system connects an instructor and a small group of students (up to 10/12 depending on the hardware quality and the instructor's bandwidth, expecially upstream).

ReBooth WebRTC connections are peer-to-peer and not peer-to-server (no SFU/MCU). There's a single audio/video WebRTC connection between the instructor and each student (the connection topology is star shaped). The instructor can individually talk with each student ("talk to booth") or communicate one-way with the whole class (class call mode). In _class call_ mode the instructor can talk to a booth, whose audio is broadcasted to the other booths (each booth can talk at a time). Students can communicate each other only via chat, with text messages transparently routed by the instructor using the WebRTC data channel. Media and binary files, on the other hand, are exchanged using a web server.

To reduce the risk of compromising student activity, ReBooth:
1) Sends the entire media file to the student's browser before the instructor starts the session or exam. This allows the student to listen to the media file in its original quality, without being affected by any connection problems or drops.
2) Records the student's audio with two separate procedures: on the one hand, it records the audio streaming the instructor receives on their computer and, on the other, it records the student's audio locally on their computer and then sends this audio to the server where ReBooth is hosted.
3) Students can save their recordings and eventually send them to the instructor by other means.

## Prerequisites

- A working [PeerJS server](https://github.com/peers/peerjs-server). See [README-PEERJS.md](/README-PEERJS.md) for a hint.
- One or more STUN servers; you may use a free existing one or setup your own (e.g. [CoTURN](https://github.com/coturn/coturn)).
- One or more TURN servers; you may buy access to an existing one or set up your own (e.g. [CoTURN](https://github.com/coturn/coturn)). You may try without a TURN server but users behind a symmetrical NAT (and in general depending on their router and firewall setup) won't be able to use ReBooth.
- A web server supporting PHP (≥ 5.5, with modules php-zip and php-mbstring enabled), HTTPS and URL rewriting to host your ReBooth installation. Here I refer to Apache but any other modern web server should be fine.
- A writeable directory on this server to store recordings and temporary data.
- A SMTP server to send invitations (optional but recommended  ).
- A PHP authentication script to restrict access only to users allowed by your organization: you'll likely have to write your own.


Having a [PeerJS server](https://github.com/peers/peerjs-server) is mandatory. The PeerJS server lets users find each other by keeping track of their connections. It basically just stores IP addresses of connected users and a hash of their usernames. 
No peer-to-peer data is routed through the server, which acts only as a connection broker. Therefore it generates very low network traffic and workload for your hardware and infrastructure.
The authors of PeerJS offer free access to their PeerJS server, but many people use it and you may find it busy: you'll likely need to deploy your own. See [README-PEERJS.md](/README-PEERJS.md) for a quick start guide.

STUN, TURN (and ICE) are a set of IETF standard protocols for negotiating traversing NATs when establishing peer-to-peer communication sessions. WebRTC and other VoIP stacks implement support for ICE to improve the reliability of IP communications. The PeerJS library used by ReBooth makes it's ICE (Interactive Connectivity Establishment) implementation by coordinating STUN and TURN to make a connection between hosts. A host uses Session Traversal Utilities for NAT (STUN) to discover its public IP address when it is located behind a NAT/Firewall. When this host wants to receive an incoming connection from another party, it provides this public IP address as a possible location where it can receive a connection. If the NAT/Firewall still won't allow the two hosts to connect directly, they make a connection to a server implementing Traversal Using Relay around NAT (TURN), which will relay media between the two parties. 

A STUN server may be enough for testing purposes but you'll definitely also need a TURN server in a production environment.

You can find plenty of free STUN servers. Google provides at least a dozen for free and a few of them are already configured in `config.js` (see below). A TURN server, on the other hand, may have a heavy bandwidth impact on your infrastructure. You may find a free one but it will likely have strong bandwidth restrictions, making it a non viable solution. You need either a commercial TURN service or to set up your own. [CoTURN](https://github.com/coturn/coturn) is an excellent open source TURN (and STUN) implementation, available on most linux distributions.

If you setup your own STUN, TURN or PeerJS server, be sure that their ports are not blocked by your firewall for inbound connections (expecially PeerJS which may not use a standard port). Also, verify that you can access those ports from your location.

Once you have your working PeerJS server and access to at least a STUN and a TURN server, you can download the ReBooth webroot from GitHub, set up the configuration files and have your ReBooth installation served by a webserver.

## Installation

Create a webroot directory and deploy rebooth into it.
Create the rebooth data directory and give the web server write permissions onto it.

```bash
[you@localhost ~]$ sudo mkdir /var/www/virtualhosts/your_site_webroot
[you@localhost ~]$ cd /var/www/virtualhosts/your_site_webroot
[you@localhost ~]$ sudo git clone https://github.com/bilo1967/rebooth.git .
[you@localhost ~]$ sudo mkdir /var/www/data/your_data_dir
[you@localhost ~]$ sudo chown www-data /var/www/data/your_data_dir
```

Then configure your web server so that it will serve your rebooth webroot directory as https://your.site/ (or whatever).

URL rewriting must be also enabled so that the extension for php scripts and html pages is removed. See [.htaccess](.htaccess) file.


## Upgrading

Basically you need to get the new version of ReBooth from GitHub (just do a ```git clone https://github.com/bilo1967/rebooth.git``` in a temporary directory) and replace everything on your webroot except your configuration files (always check for additions or changes), authentication scripts, and any images you may have replaced or customized. Always make a backup copy of your webroot before trying an upgrade.


## Configuration and authentication module

There are two configuration files you have to set up:
- [config/config.js](config/config.js) for the javascript code running on your browser
- [config/config.inc.php](config/config.inc.php) for the PHP scripts running on the server

You will also need to write your PHP own authentication module.

### config.js

Here you basically just have to configure the javascript access to your PeerJS, STUN and TURN servers.
```js
const PeerJSConfig = {
    host:   'your.peerjs.server', // your peerjs server address
    port:   9443,                 // your peerjs server port
    path:   '/',                  // your peerjs path
    secure: true,                 // your peerjs server uses SSL?
    
    // You may use a STUN-only configuration for testing purposes but if you need
    // to actually reach (mostly) everyone you definitely need also a TURN server...
    config: {
        iceServers: [
            { urls: 'stun:stun.l.google.com:19302' },
            { urls: 'stun:stun1.l.google.com:19302' },
            { urls: 'stun:stun2.l.google.com:19302' },
//          { urls: 'turn:numb.viagenie.ca', username: 'webrtc@live.com', credential: 'muazkh'},
//          { urls: 'turn:0.peerjs.com:3478', username: 'peerjs', credential: 'peerjsp' },
        ]
    },
};
```
At the moment you can leave any other setting as is.

In the provided configuration file there are public STUN servers but no TURN server so, depending on user's router and firewall setup, it might not always be possible to connect to ReBooth.


### config.inc.php

ReBooth needs:
* a SMPT server to send email invitations to students;
* a few info to correctly generate these invitations;
* a directory writable by the web server to store instructor's media files, student's recordings and temporary stuff;
* private and public key for Google Captcha v2 (optional);
* instructions on how to access your authentication mudule (see next section).

All of these settings must be implemented in the [config/config.inc.php](config/config.inc.php) file. See the comments on the file itself for details.


### Authentication module

_ReBooth_ is meant to be used within your organization and to authenticate users via your infrastructure. Some examples are provided but no authentication mechanism is actually implemented: you will have to write your authentication module by your own. In this module you will implement an authentication function accepting a username and a password as parameters which returns _true_ if the authentication succeeds, _false_ otherwise. The module file has to be placed be in the [auth](auth) directory, where you'll find a few examples that you may adapt to your needs. The authentication module file and function name are then set in the [config/config.inc.php](config/config.inc.php) file. 

This is an example dummy authentication module accepting any valid email address as username and 'apriti sesamo' as password.
```php
<?php
    function dummyAuth($user, $password) {
        return (filter_var($user, FILTER_VALIDATE_EMAIL) && $password == 'apriti sesamo');
    }
?>
```

In the [auth directory](/auth) there are a couple of example modules with radius and imap authentication. You may adapt them to fit your needs by setting the value of a few constants at the beginning of each module.

There's also a [guest-auth.inc.php](auth/guest-auth.inc.php) authentication module wich implements a basic local authentication system. Credentials are retrieved from a JSON file which is stored into the data directory and managed using a command line [helper php script located in the bin directory](bin/guest.php). Run the script without parameters to get help. Users can be added, deleted, modified. For each of them it is possible to set a username, a password and an account expiry (in days).

In this case you would set something like this into your config.inc.php file:

```php
    "auth_module"               => 'guest-auth.inc.php',
    "auth_function"             => 'guestAuth',
    "auth_description"          => 'Insert your credentials to log in',
    "auth_username_placeholder" => 'type your username',
    "auth_password_placeholder" => 'type your password',
```

You may combine modules to implement a cascade authentication system. See [auth/local-auth.inc.php](auth/local-auth.inc.php) as an example.


## License

ReBooth is available under the [GNU Affero General Public v3](https://www.gnu.org/licenses/agpl-3.0.html) License. 

In brief, this program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as  published by the Free Software Foundation, either version 3 of the  License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but **WITHOUT ANY WARRANTY**; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the [GNU Affero General Public License](https://www.gnu.org/licenses/agpl-3.0.html) for more details.

Documentation is available under [CC BY 3.0](https://creativecommons.org/licenses/by/3.0/) license. 

ReBooth uses several third-party frameworks, components, libraries and resources:
* [PeerJS](https://peerjs.com/">https://peerjs.com/)
* [JQuery](https://jquery.com/">https://jquery.com/)
* [Bootstrap 4](https://getbootstrap.com/)
* [Toastr] (https://codeseven.github.io/toastr/)
* [BootBox JS](http://bootboxjs.com/)
* [JsZip](https://stuk.github.io/jszip/)
* [Animate.css](https://github.com/animate-css/animate.css)
* [FontAwesome](https://fontawesome.com/)
* [Code snippets from WebRTC Samples](https://webrtc.github.io/samples/)
* [PHP Mailer](https://github.com/PHPMailer/PHPMailer)
* [JSCookie] (https://github.com/js-cookie/js-cookie)
* [Moment JS](https://momentjs.com/)


CHANGELOG.md
## [v0.8.4]
Improvements to the chat system (the most visible is the emoji picker).  
More tooltips have been added.  
Minor code cleanup.  
Removed unneeded javascript library.

## [v0.8.3]
Many minor fixes.
File manager content syncronization between concurrent sessions.  
Regression fixed on session timeout handling.

## [v0.8.2]
Enhanced session timeout 

## [v0.8.1]
Debug messages can now be logged to server (no need to collect them from teacher and booths)

## [v0.8.0]
In "class call" mode it is now possible to talk to a single booth and to broadcast its audio to the others. A single booth can talk at a time but everybody now can listen to it.

## [v0.7.3]
Class call mode is now displayed also on each single booth.  
When a booth connects after the class has actually started now can see if the class call mode is active or if the teacher is talking to someone.

## [v0.7.2]
The vumeter code used to visually check if the microphone is working has been rewrittend and added to the instructor setup panel, also.

## [v0.7.1]
Chat display code has been rewritten.  
Minor bug fixes.  
Minor file manager interface changes.

## [v0.7.0]
Introduced video media support.

## [v0.6.6]
Audio feedback when sims/cons are over.

## [v0.6.5]
Source audio overboost option in configuration panel for booths.

## [v0.6.4]
Introduced optional "double track" recording on booth's recorder (player audio track on R channel, booth audio on L channel). This feature is disabled by default and can be enabled (at run time) in the audio device configuration panel by the instructor.

## [v0.6.3]
Added booths volume slider to the instructor interface.  
Volume overboost option for audio file player for instructor and booth.  
Volume overboost option for booth audio for instructor.
Added credits page.

## [v0.6.2]
Introduced the "split audio" feature. The instructor can now hear booth on one channel of his/her headphone and the floor audio on the other. This feature is disabled by default and can be enabled (at run time) in the audio device configuration panel by the instructor.

## [v0.6.1]
Initial GitHub release


config/config.js
// The only thing you need to change is the peerjs server (PeerJSConfig constant below)
// Avoid editing the rest if you don't know what you are doing

const DebugMode = false;
const DebugToServer = false;
const DebugLevel = 1; // 0=errors, 1=warnings, 2=info, 3=details
const BetaFeatures = false;

const PinLength = 6;

const SimultaneousBatchDefaultDelay = 15;
const ConsecutiveBatchDefaultDelay  = 0;

// Maximum gain values (1=100%)
// WARNING: overboosting audio volume may cause damage to your hearing
const BoothsGainMax  = 2.25; // Boost booths volume (instructor only)
const PlayerGainMax  = 1.75; // Boost player volume (both instructor and booth)
const TeacherGainMax = 1.00; // Boost teacher volume (unused)

// Booth check interval
const BoothPollingInterval = 3210;


// Allowed uploadable media files
const AllowedMediaFileExtensions = ['wav','mp3','mpeg3','ogg','m4a','mp4','m4v'];

// Prefix to use to name booths added at class time (will prefix a number)
const classTimeBoothNamePrefix = "student@booth.n";


// Teacher PC has to handle several upstream and downstream connections
// Tipically downstream band is larger than upstream so it make sense to
// limit upstream bandwidth 
// Student PC has to handle just one upstream and one downstream. We can
// allow a little more upstream bandwidth than the Teacher since the 
// Teacher downstream width is larger than upstream
const BandwidthProfiles = {
    max: { 
        // Teacher: 256kbps upload @240p
        TeacherVideoBW:   192000,
        TeacherAudioBW:   64000,
        TeacherVRes:      240,
        // Student: 332kbps upload @360p
        StudentVideoBW:   236000,
        StudentAudioBW:   96000,
        StudentVRes:      480
    },    
    high: {
        // Teacher: 224kbps upload @160p
        TeacherVideoBW:   160000,
        TeacherAudioBW:   64000,
        TeacherVRes:      160,
        // Student: 288kbps upload @240p
        StudentVideoBW:   192000,
        StudentAudioBW:   96000,
        StudentVRes:      240
    },
    medium: {
        // Teacher: 192kbps upload @120p
        TeacherVideoBW:   136000,
        TeacherAudioBW:   56000,
        TeacherVRes:      120,
        // Student: 256kbps upload @160p
        StudentVideoBW:   162000,
        StudentAudioBW:   94000,
        StudentVRes:      160
    },
    low: {
        // Teacher: 150kbps upload @120p
        TeacherVideoBW:   96000,
        TeacherAudioBW:   54000,
        TeacherVRes:      120,
        // Teacher: 200kbps upload @120p
        StudentVideoBW:   110000,
        StudentAudioBW:   90000,
        StudentVRes:      120
    },
    default: null
};
BandwidthProfiles.default = BandwidthProfiles.low;

const LocalCameraConstraints = {
/*    
    video: {
        frameRate: {ideal: 15, max: 30},
//      width: {ideal: 160 },
        height: {ideal: 120, max: 480 }
    },
*/    
    video: {
        frameRate: {ideal: 15, max: 30},
    },
    audio: true,
/*    
    audio: {
        echoCancellation: false,
        noiseSuppression: false,
    }
*/
};


const ScreenShareConstraints = {
    video: {
//      cursor: 'always',
//      frameRate: 15,
        height: { ideal: 480, max: 720 },
    },
    audio: {
        echoCancellation: true,
        noiseSuppression: true,
        sampleSize: 8,
        sampleRate: 22050,
        channelCount: 1    
    }
};

const PeerJSConfig = {
    host:   'your.peerjs.server', // your peerjs server address
    port:   9443,                 // your peerjs server port
    debug:  2,
    path:   '/',
    secure: true,                 // use SSL
    
    // You may use a STUN-only configuration for testing purposes but if you need
    // to actually reach (mostly) everyone you definitely need also a TURN server...
    config: {
        iceServers: [
            { urls: 'stun:stun.l.google.com:19302' },
            { urls: 'stun:stun1.l.google.com:19302' },
            { urls: 'stun:stun2.l.google.com:19302' },
//          { urls: 'turn:numb.viagenie.ca', username: 'webrtc@live.com', credential: 'muazkh'},
//          { urls: 'turn:0.peerjs.com:3478', username: 'peerjs', credential: 'peerjsp' },
        ]
    },
};
